### PR TITLE
Publisher#flatMap*DelayError limit queued exceptions

### PR DESCRIPTION
**Pull request by Scottmitch**
Motivation:
Publisher#flatMap*DelayError operators will queue exceptions and add
them all as suppressed exceptions once the upstream source and all
mapped publishers have terminated. However this operator maybe used for
unbounded streams, with retry operators, and this may result in a large
amount of exceptions which will exhaust memory over time.

Modifications:
- Limit the number of exception that is accumulated, and allow users to
override the default value.

Result:
Publisher#flatMap*DelayError bounds the number of accumulated exceptions
and will not exhaust memory for infinite streams.

### CODE REVIEW

**Comment by Scottmitch** [comment]
Merge after https://github.com/apple/servicetalk/pull/1218

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java [hunk 1]
```diff
@@ -383,6 +393,7 @@ private void drainPending() {
                                     ++emittedCount;
                                 }
                             } catch (Throwable cause) {
+                                ++emittedCount;
```
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java:396]: Why this change is required?

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java [hunk 2]
```diff
@@ -68,26 +69,35 @@
...
+
+    PublisherFlatMapMerge(Publisher<T> original, Function<? super T, ? extends Publisher<? extends R>> mapper,
+                          int maxDelayedErrors, int maxConcurrency, Executor executor) {
         super(original, executor);
-        this.mapper = requireNonNull(mapper);
         if (maxConcurrency <= 0) {
             throw new IllegalArgumentException("maxConcurrency: " + maxConcurrency + " (expected >0)");
         }
+        if (maxDelayedErrors < 0) {
+            throw new IllegalArgumentException("maxConcurrency: " + maxDelayedErrors + " (expected >=0)");
```
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java:96]: `maxConcurrency` -> `maxDelayedErrors`

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java [hunk 3]
```diff
@@ -588,7 +599,7 @@ public void onNext(@Nullable final R r) {
 
             @Override
             public void onError(final Throwable t) {
-                if (!parent.source.delayError) {
+                if (parent.source.maxDelayedErrors <= 0) {
```
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java:602]: Values < 0 are not allowed, consider using `parent.source.maxDelayedErrors == 0` to make it more clear what `0` value means.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java [hunk 4]
```diff
@@ -539,7 +589,62 @@ protected Publisher() {
...
+     * @param maxDelayedErrorsHint The maximum amount of errors that will be queued. After this point exceptions maybe
+     * discarded to reduce memory consumption.
+     * @param <R> Type of items emitted by the returned {@link Publisher}.
+     * @return A new {@link Publisher} that emits all items emitted by each single produced by {@code mapper}.
+     *
+     * @see <a href="http://reactivex.io/documentation/operators/merge.html">ReactiveX merge operator.</a>
+     */
+    public final <R> Publisher<R> flatMapMergeSingleDelayError(
+            Function<? super T, ? extends Single<? extends R>> mapper, int maxConcurrency, int maxDelayedErrorsHint) {
+        return new PublisherFlatMapSingle<>(this, mapper, maxDelayedErrorsHint, maxConcurrency, executor);
```
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java:647]: `maxDelayedErrorsHint == 0` has a special different meaning. Users may think that if they use `*DelayError` methods and pass `0` here, all exceptions will be delayed, but none of them will be accumulated in the `CompositeException`. However, they will start seeing exceptions asap.

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapSingle.java [hunk 5]
```diff
@@ -56,25 +58,32 @@
...
     PublisherFlatMapSingle(Publisher<T> original, Function<? super T, ? extends Single<? extends R>> mapper,
-                           int maxConcurrency, boolean delayError, Executor executor) {
+                           int maxDelayedErrors, int maxConcurrency, Executor executor) {
         super(original, executor);
-        this.mapper = requireNonNull(mapper);
         if (maxConcurrency <= 0) {
             throw new IllegalArgumentException("maxConcurrency: " + maxConcurrency + " (expected > 0)");
         }
+        if (maxDelayedErrors < 0) {
+            throw new IllegalArgumentException("maxConcurrency: " + maxDelayedErrors + " (expected >=0)");
```
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapSingle.java:82]: `maxConcurrency` -> `maxDelayedErrors`

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapSingle.java [hunk 6]
```diff
@@ -390,12 +399,12 @@ public void onSuccess(@Nullable R result) {
 
             @Override
             public void onError(Throwable t) {
-                if (!source.delayError) {
+                if (source.maxDelayedErrors <= 0) {
```
Comment (idelpivnitskiy) [inline servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapSingle.java:402]: Values < 0 are not allowed, consider using `parent.source.maxDelayedErrors == 0` to make it more clear what `0` value means.

**Review by idelpivnitskiy — COMMENTED** [review]
Good idea, a few comments:

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java (hunk 1 above)
Comment (Scottmitch) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/PublisherFlatMapMerge.java:396]: if an exception is caught here it is bcz an item has been delivered and the subscriber threw, so we need decrement the count (missed previously).

File: servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java (hunk 4 above)
Comment (Scottmitch) [reply servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/Publisher.java:647]: good point, `maxDelayedErrorsHint == 0` shouldn't be allowed here.

**Comment by Scottmitch** [comment]
ci failure...
```
19:09:42 Creating servicetalk-java8-prb_runtime-setup_1 ... 
19:09:44 
Creating servicetalk-java8-prb_runtime-setup_1 ... done
Error: No such container: 587f943d10e0fb045a3cc441e20f8774dd679dbe27ba850c011b3ac068b4247c
19:09:44 No such container: 587f943d10e0fb045a3cc441e20f8774dd679dbe27ba850c011b3ac068b4247c
19:09:44 Build step 'Execute shell' marked build as failure
```

test this please

**Review by idelpivnitskiy — APPROVED** [review]
🚀
